File: packages/core/src/rule.ts
import type { EventType, Rule, RuleMode } from './types.js'

let ruleCounter = 0

export interface RuleConfig {
  name?: string
  triggers: EventType[]
  mode: RuleMode
  guard?: (...payloads: any[]) => boolean
  action: (...payloads: any[]) => any
  outputs: EventType[]
  priority?: number
}

/**
 * Create a Rule node for the DAG.
 */
export function createRule(config: RuleConfig): Rule {
  const id = `rule_${ruleCounter++}`
  const isJoin = config.mode === 'join'
  const rule: Rule = {
    id,
    name: config.name ?? id,
    triggers: config.triggers,
    mode: config.mode,
    guard: config.guard,
    action: config.action,
    outputs: config.outputs,
    priority: config.priority ?? 0,
    _disposed: false,
    _joinPending: isJoin ? config.triggers.map(() => []) : null,
    _joinReady: 0,
  }
  return rule
}

/**
 * Register a rule as a consumer of its trigger event types.
 */
export function registerRuleConsumers(rule: Rule): void {
  for (const trigger of rule.triggers) {
    trigger._consumers.add(rule)
    // Cache solo consumer for fast-path dispatch
    trigger._solo = trigger._consumers.size === 1 ? rule : null
  }
}

/**
 * Unregister a rule from its trigger event types.
 */
export function unregisterRuleConsumers(rule: Rule): void {
  for (const trigger of rule.triggers) {
    trigger._consumers.delete(rule)
    if (trigger._consumers.size === 1) {
      trigger._solo = trigger._consumers.values().next().value ?? null
    } else {
      trigger._solo = null
    }
  }
  rule._disposed = true
}

/** Reset counter (for testing) */
export function resetRuleCounter(): void {
  ruleCounter = 0
}

File: packages/dart/lib/src/rule.dart
import 'event_type.dart';

/// Rule mode: 'each' fires per-event, 'join' fires when all input types ready.
enum RuleMode { each, join }

int _ruleCounter = 0;

/// A node in the DAG representing a reactive rule.
class Rule {
  final String id;
  final String name;
  final List<EventType> triggers;
  final RuleMode mode;
  final bool Function(List<dynamic>)? guard;
  final dynamic Function(List<dynamic>) action;
  final List<EventType> outputs;
  final int priority;
  bool disposed = false;

  Rule._({
    required this.id,
    required this.name,
    required this.triggers,
    required this.mode,
    this.guard,
    required this.action,
    required this.outputs,
    required this.priority,
  });

  @override
  String toString() => 'Rule($name)';
}

/// Configuration for creating a [Rule].
class RuleConfig {
  final String? name;
  final List<EventType> triggers;
  final RuleMode mode;
  final bool Function(List<dynamic>)? guard;
  final dynamic Function(List<dynamic>) action;
  final List<EventType> outputs;
  final int priority;

  const RuleConfig({
    this.name,
    required this.triggers,
    required this.mode,
    this.guard,
    required this.action,
    required this.outputs,
    this.priority = 0,
  });
}

/// Create a [Rule] node for the DAG.
Rule createRule(RuleConfig config) {
  final id = 'rule_${_ruleCounter++}';
  return Rule._(
    id: id,
    name: config.name ?? id,
    triggers: config.triggers,
    mode: config.mode,
    guard: config.guard,
    action: config.action,
    outputs: config.outputs,
    priority: config.priority,
  );
}

/// Register a rule as a consumer of its trigger event types.
void registerRuleConsumers(Rule rule) {
  for (final trigger in rule.triggers) {
    trigger.consumers.add(rule);
  }
}

/// Unregister a rule from its trigger event types.
void unregisterRuleConsumers(Rule rule) {
  for (final trigger in rule.triggers) {
    trigger.consumers.remove(rule);
  }
  rule.disposed = true;
}

/// Reset counter (for testing).
void resetRuleCounter() {
  _ruleCounter = 0;
}
